collector/collector.go

```diff
@@ -1,159 +1,131 @@
-type Collector struct {
-	Enabled       *bool
-	Name          string
-	PromCollector prometheus.Collector
-	Errors        CollectErrors
-	CLIHelper     CLIHelper
-	Logger        log.Logger
-}
+func registerCollector(name string, enabledByDefaultStatus bool, factory func(logger log.Logger) (Collector, error)) {
+	defaultState := "disabled"
+	if enabledByDefaultStatus {
+		defaultState = "enabled"
+	}
 
-// NewExporter returns an Exporters type containing a slice of Collectors.
-func NewExporter(collectors []*Collector) *Exporters {
-	return &Exporters{Collectors: collectors}
+	factories[name] = factory
+	collectorState[name] = kingpin.Flag(fmt.Sprintf("collector.%s", name), fmt.Sprintf("Enable the %s collector (default: %s).", name, defaultState)).Default(strconv.FormatBool(enabledByDefaultStatus)).Bool()
 }
 
-// SetVTYSHPath sets the path of vtysh.
-func (e *Exporters) SetVTYSHPath(path string) {
-	vtyshPath = path
+// Collector is the interface a collector has to implement.
+type Collector interface {
+	// Update metrics and sends to the Prometheus.Metric channel.
+	Update(ch chan<- prometheus.Metric) error
 }
 
-// SetVTYSHTimeout sets the path of vtysh.
-func (e *Exporters) SetVTYSHTimeout(timeout time.Duration) {
-	vtyshTimeout = timeout
+// Exporter collects all collector metrics, implemented as per the prometheus.Collector interface.
+type Exporter struct {
+	Collectors map[string]Collector
+	logger     log.Logger
 }
 
-// SetVTYSHOptions sets the options passed to vtysh
-func (e *Exporters) SetVTYSHOptions(s string) {
-	frrVTYSHOptions = s
-}
+// NewExporter returns a new Exporter.
+func NewExporter(logger log.Logger) (*Exporter, error) {
+	collectors := make(map[string]Collector)
 
-// SetVTYSHSudo sets the first command to execute vtysh if sudo is enabled.
-func (e *Exporters) SetVTYSHSudo(enable bool) {
-	vtyshSudo = enable
-}
+	initiatedCollectorsMtx.Lock()
+	defer initiatedCollectorsMtx.Unlock()
 
-// Describe implemented as per the prometheus.Collector interface.
-func (e *Exporters) Describe(ch chan<- *prometheus.Desc) {
-	for _, desc := range frrDesc {
-		ch <- desc
-	}
-	for _, collector := range e.Collectors {
-		collector.PromCollector.Describe(ch)
+	for name, enabled := range collectorState {
+		if !*enabled {
+			continue
+		}
+		if collector, exists := initiatedCollectors[name]; exists {
+			collectors[name] = collector
+		} else {
+			collector, err := factories[name](log.With(logger, "collector", name))
+			if err != nil {
+				return nil, err
+			}
+			collectors[name] = collector
+			initiatedCollectors[name] = collector
+		}
 	}
+	return &Exporter{
+		Collectors: collectors,
+		logger:     logger,
+	}, nil
 }
 
 // Collect implemented as per the prometheus.Collector interface.
-func (e *Exporters) Collect(ch chan<- prometheus.Metric) {
+func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
+	frrTotalScrapeCountMtx.Lock()
 	frrTotalScrapeCount++
+	frrTotalScrapeCountMtx.Unlock()
```

Comment:
Consider replacing this with a [prometheus.Counter](https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#Counter), which are concurrency-safe and don't require you to implement your own locking.

Reply:
Good call. Done.